dev/NavigationView/NavigationView.cpp

```diff
@@ -379,10 +427,354 @@ void NavigationView::OnApplyTemplate()
+        {
+            winrt::IndexPath ip = GetIndexPathForContainer(nvib);
+            m_selectionModel.SelectAt(ip);
+        }
+        else if (!newIsSelected && newItemIsSelectedItem)
+        {
+            auto indexPath = GetIndexPathForContainer(nvib);
+            auto indexPathFromModel = m_selectionModel.SelectedIndex();
+
+            if (indexPath && indexPathFromModel && indexPath.GetAt(0) == indexPathFromModel.GetAt(0))
+            {
+                m_selectionModel.DeselectAt(indexPath);
+            }
+        }
+    }
+}
+
+void NavigationView::RaiseItemInvokedForNavigationViewItem(const winrt::NavigationViewItem& nvi)
+{
+    winrt::IInspectable nextItem = nullptr;
+    auto prevItem = SelectedItem();
+    auto parentIR = GetParentItemsRepeaterForContainer(nvi);
+
+    bool isInOverflow = parentIR.Name() == c_overflowRepeater;
+    bool itemSelectsOnInvoked = nvi.SelectsOnInvoked();
+
+    auto itemIndex = parentIR.GetElementIndex(nvi);
+    auto itemsSourceView = parentIR.ItemsSourceView();
+    if (itemsSourceView)
+    {
+        auto inspectingDataSource = static_cast<InspectingDataSource*>(winrt::get_self<ItemsSourceView>(itemsSourceView));
+        nextItem = inspectingDataSource->GetAt(itemIndex);
+    }
+
+    // Other transition other than default only apply to topnav
+    // when clicking overflow on topnav, transition is from bottom
+    // otherwise if prevItem is on left side of nextActualItem, transition is from left
+    //           if prevItem is on right side of nextActualItem, transition is from right
+    // click on Settings item is considered Default
+    NavigationRecommendedTransitionDirection recommendedDirection = NavigationRecommendedTransitionDirection::Default;
+    if (IsTopNavigationView() && itemSelectsOnInvoked)
+    {
+        if (isInOverflow)
+        {
+            recommendedDirection = NavigationRecommendedTransitionDirection::FromOverflow;
+        }
+        else if (prevItem && nvi)
+        {
+            recommendedDirection = GetRecommendedTransitionDirection(NavigationViewItemBaseOrSettingsContentFromData(prevItem), nvi);
+        }
+    }
+
+    RaiseItemInvoked(nextItem, false /*isSettings*/, nvi, recommendedDirection);
+}
+
+void NavigationView::OnNavigationViewItemInvoked(const winrt::NavigationViewItem& nvi)
+{
+    auto selectedItem = SelectedItem();
+    RaiseItemInvokedForNavigationViewItem(nvi);
+    auto selectedItemAfterInvokeCallback = SelectedItem();
+
+    // User changed selectionstate in the ItemInvoked callback
+    if (selectedItem != selectedItemAfterInvokeCallback)
+    {
+        return;
+    }
+
+    bool itemSelectsOnInvoked = nvi.SelectsOnInvoked();
+    if (m_selectionModel && itemSelectsOnInvoked)
+    {
+        winrt::IndexPath ip = GetIndexPathForContainer(nvi);
+        m_selectionModel.SelectAt(ip);
+    }
+
+    ClosePaneIfNeccessaryAfterItemIsClicked();
+}
+
+bool NavigationView::IsRootItemsRepeater(winrt::hstring name)
```

Comment:
Perhaps this method could just take a framework element or even UIElement instead of the name. That way in all the callsites, you don't have to do the try_as and also check for null.